OpenNefia.Content/UI/Element/FeatWindow.cs

```diff
@@ -0,0 +1,311 @@
+        private IAssetDrawable AssetDecoFeatB;
+        private IAssetDrawable AssetDecoFeatC;
+        private IAssetDrawable AssetDecoFeatD;
+
+        private Func<Dictionary<FeatPrototype, int>> GetFeatsFunc;
+        private Action<FeatData.Feat> SelectFeatAction;
+        private Func<int> GetAvailableCountFunc;
+
+        public FeatWindow(Func<Dictionary<FeatPrototype, int>> getFeatsFunc, Action<FeatData.Feat> selectFeatAction, Func<int> getAvailableCountFunc) 
+            : base(keyHintXOffset: 64)
+        {
+            _prototypeManager = IoCManager.Resolve<IPrototypeManager>();
+            AssetInventoryIcons = InventoryHelpers.MakeIcon(InventoryIcon.Feat);
+            AssetDecoFeatA = new AssetDrawable(Protos.Asset.DecoFeatA);
+            AssetDecoFeatB = new AssetDrawable(Protos.Asset.DecoFeatB);
+            AssetDecoFeatC = new AssetDrawable(Protos.Asset.DecoFeatC);
+            AssetDecoFeatD = new AssetDrawable(Protos.Asset.DecoFeatD);
+
+            NameTopic = new UiTextTopic(Loc.GetString("Elona.FeatMenu.NameTopic"));
+            DetailTopic = new UiTextTopic(Loc.GetString("Elona.FeatMenu.DetailTopic"));
+            FeatCountText = new UiText(UiFonts.WindowPage);
+
+            TextTitle.Text = Loc.GetString("Elona.FeatMenu.Title");
+            List = new UiPagedList<FeatData>(ItemsPerPage, this, new Vector2i(-55, -3));
+
+            GetFeatsFunc = getFeatsFunc ?? (() => new Dictionary<FeatPrototype, int>());
+            SelectFeatAction = selectFeatAction ?? (feat => { });
+            GetAvailableCountFunc = getAvailableCountFunc ?? (() => 0);
+
+            List.GrabFocus();
+            List.EventOnActivate += List_OnActivate;
+            RefreshData();
+        }
+
+        private void RefreshData(FeatPrototype? lastSelected = null)
+        {
+            List.Clear();
+            
+            var data = new List<FeatCell>();
+            var prototypes = _prototypeManager.EnumeratePrototypes<FeatPrototype>()?.Where(x => x.FeatType == FeatType.Feat)!;
+            var featCount = GetAvailableCountFunc();
+            var gainedFeats = GetFeatsFunc();
+
+            if (featCount > 0)
+            {
+                data.Add(new FeatCell(new FeatData.FeatHeader("Elona.FeatMenu.AvailableHeader")));
+                data.AddRange(prototypes.Select(x =>
+                {
+                    gainedFeats.TryGetValue(x, out var val);
+                    return new FeatCell(new FeatData.Feat(x, val));
+                }));
+                FeatCountText.Text = string.Format(Loc.GetString("Elona.FeatMenu.FeatCount"), featCount);
+            }
+            else
+            {
+                FeatCountText.Text = string.Empty;
+            }
+            data.Add(new FeatCell(new FeatData.FeatHeader("Elona.FeatMenu.GainedHeader")));
+            data.AddRange(gainedFeats
+                .Select(x => new FeatCell(new FeatData.GainedFeat(x.Key, x.Value, x.Key.FeatType == FeatType.Race)))
```

Comment:
Is there a reason this is only filtering by race? And why the ordering is like this?

Reply:
it isn't filtering here, just setting the IsRace variable of the record

Reply:
changed in b1407d8
made it actually just sort based on the FeatType, the field was just redundant